pkg/bitlpm/cidr.go

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package bitlpm
+
+import (
+	"math/bits"
+	"net/netip"
+)
+
+// CIDRTrie can hold both IPv4 and IPv6 prefixes
+// at the same time.
+type CIDRTrie[T any] struct {
+	v4 Trie[Key[netip.Prefix], T]
+	v6 Trie[Key[netip.Prefix], T]
+}
+
+// NewCIDRTrie creates a new CIDRTrie[T any].
+func NewCIDRTrie[T any]() *CIDRTrie[T] {
+	return &CIDRTrie[T]{
+		v4: NewTrie[netip.Prefix, T](32),
+		v6: NewTrie[netip.Prefix, T](128),
+	}
+}
+
+// Lookup returns the longest matched value for a given address.
+func (c *CIDRTrie[T]) Lookup(addr netip.Addr) T {
+	tri := c.v4
+	bits := 32
+	if addr.Is6() {
+		tri = c.v6
+		bits = 128
+	}
+	var def T
+	tri.Ancestors(uint(bits), cidrKey(netip.PrefixFrom(addr, bits)), func(_ uint, _ Key[netip.Prefix], v T) bool {
+		def = v
+		return true
+	})
+	return def
```

Comment:
nit: Could this do `tri.Lookup()`?
```suggestion
	return tri.Lookup(cidrKey(netip.PrefixFrom(addr, bits)))
```

Reply:
Or, I guess, for more code reuse:

```
func (c *CIDRTrie[T]) Lookup(addr netip.Addr) T {
        bits := addr.BitLen()
        prefix := netip.PrefixFrom(addr, bits)
        tri := c.treeForFamily(prefix)
        return tri.Lookup(cidrKey(prefix))
}
```